blob/s3blob/s3blob.go

```diff
@@ -377,43 +364,67 @@ func (b *bucket) ListPaged(ctx context.Context, opts *driver.ListOptions) (*driv
 	return &page, nil
 }
 
-func (b *bucket) listObjectsV2Request(in *s3.ListObjectsV2Input) (req *request.Request, output *s3.ListObjectsV2Output) {
-	if b.useLegacyList {
-		marker := in.ContinuationToken
-		if marker == nil || (marker != nil && in.StartAfter != nil && *in.StartAfter > *marker) {
-			marker = in.StartAfter
+func (b *bucket) listObjects(in *s3.ListObjectsV2Input, opts *driver.ListOptions) (*s3.ListObjectsV2Output, error) {
+	if !b.useLegacyList {
+		if opts.BeforeList != nil {
+			asFunc := func(i interface{}) bool {
+				p, ok := i.(**s3.ListObjectsV2Input)
+				if !ok {
+					return false
+				}
+				*p = in
+				return true
+			}
+			if err := opts.BeforeList(asFunc); err != nil {
+				return nil, err
+			}
 		}
-
-		legacyIn := &s3.ListObjectsInput{
-			Bucket:       in.Bucket,
-			Delimiter:    in.Delimiter,
-			EncodingType: in.EncodingType,
-			Marker:       marker,
-			MaxKeys:      in.MaxKeys,
-			Prefix:       in.Prefix,
-			RequestPayer: in.RequestPayer,
+		req, resp := b.client.ListObjectsV2Request(in)
+		if err := req.Send(); err != nil {
+			return nil, err
 		}
-		req, legacyResp := b.client.ListObjectsRequest(legacyIn)
+		return resp, nil
+	}
 
-		var nextContinuationToken *string
-		if legacyResp.NextMarker != nil && *legacyResp.NextMarker != "" {
-			nextContinuationToken = legacyResp.NextMarker
-		} else if !*legacyResp.IsTruncated {
-			nextContinuationToken = aws.String(*legacyResp.Contents[len(legacyResp.Contents)-1].Key + " ")
-		} else {
-			nextContinuationToken = nil
+	// Use the legacy ListObjects request.
+	legacyIn := &s3.ListObjectsInput{
+		Bucket:       in.Bucket,
+		Delimiter:    in.Delimiter,
+		EncodingType: in.EncodingType,
+		Marker:       in.ContinuationToken,
+		MaxKeys:      in.MaxKeys,
+		Prefix:       in.Prefix,
+		RequestPayer: in.RequestPayer,
+	}
+	if opts.BeforeList != nil {
+		asFunc := func(i interface{}) bool {
+			p, ok := i.(**s3.ListObjectsInput)
+			if !ok {
+				return false
+			}
+			*p = legacyIn
+			return true
 		}
-
-		resp := &s3.ListObjectsV2Output{
-			CommonPrefixes:        legacyResp.CommonPrefixes,
-			Contents:              legacyResp.Contents,
-			NextContinuationToken: nextContinuationToken,
+		if err := opts.BeforeList(asFunc); err != nil {
+			return nil, err
 		}
+	}
+	req, legacyResp := b.client.ListObjectsRequest(legacyIn)
+	if err := req.Send(); err != nil {
+		return nil, err
+	}
 
-		return req, resp
-	} else {
-		return b.client.ListObjectsV2Request(in)
+	var nextContinuationToken *string
+	if legacyResp.NextMarker != nil {
+		nextContinuationToken = legacyResp.NextMarker
+	} else if aws.BoolValue(legacyResp.IsTruncated) {
+		nextContinuationToken = aws.String(aws.StringValue(legacyResp.Contents[len(legacyResp.Contents)-1].Key) + " ")
```

Comment:
Why is there a `" "` added?

Reply:
It was in the PR before; I thought it was needed to ensure that the next set of results doesn't include this one. But, I tried without it and it doesn't seem to be necessary, so removed it.